File: src/dataframe/rust/stats/distributions/hypergeometric.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust hypergeometric functions...\n");

  const testCases = [
    { func: "dhyper", distribution: "hypergeometric", args: [3, 10, 5, 7, 0] },
    {
      func: "phyper",
      distribution: "hypergeometric",
      args: [3, 10, 5, 7, 1, 0],
    },
    {
      func: "qhyper",
      distribution: "hypergeometric",
      args: [0.1002331, 10, 5, 7, 1, 0],
    },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rhyper separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rhyperResult = await callR("rhyper", "hypergeometric", 1, 10, 5, 7);
    console.log(`rhyper(10, 5, 7) = ${rhyperResult}`);
  } catch (error) {
    console.log(`rhyper: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}

File: packages/dataframe/rust/stats/distributions/hypergeometric.test.R
#!/usr/bin/env Rscript

# Hypergeometric distribution test script
# Usage: Rscript hypergeometric.test.R <function> <x_or_p> <m> <n> <k> [lower_tail] [log_p]

source("test-helpers.R")

# Define hypergeometric distribution functions
hypergeometric_density <- function(x, m, n, k, log = FALSE) {
  dhyper(x, m, n, k, log = log)
}

hypergeometric_cumulative <- function(x, m, n, k, lower.tail = TRUE, log.p = FALSE) {
  phyper(x, m, n, k, lower.tail = lower.tail, log.p = log.p)
}

hypergeometric_quantile <- function(p, m, n, k, lower.tail = TRUE, log.p = FALSE) {
  qhyper(p, m, n, k, lower.tail = lower.tail, log.p = log.p)
}

hypergeometric_random <- function(nn, m, n, k) {
  rhyper(nn, m, n, k)
}

# Run the test
run_distribution_test(hypergeometric_density, hypergeometric_cumulative, hypergeometric_quantile, hypergeometric_random)